### PR TITLE
Distill megatron - test Draft WIP

### PR DESCRIPTION
**Pull request by younesbelkada**
An attempt to perform knowledge distillation using Megatron-DeepSpeed

disclaimer: this is a super ugly version of the code, the PR is here to compare the difference between the original code and this modified version - for now I don't plan to merge this PR

# Updates on 28.09.2022

This version is very ugly, I had to add an argument `student_` on all megatron modules since the arguments are directly retrieved from the global variable. The other solution could be to have each class re-written with the suffix `Student` - eg `GPTModelStudentPipe`. I preferred the first solution to have a quick working implementation. 

The forward and backward pass seems to pass for the student model - for now I am not computing the teacher's logits.. 
Two solutions for that 

1 - [In `distill_train_step`](https://github.com/younesbelkada/Megatron-DeepSpeed/blob/26d9ab0600c6afad93c11873306bad3dea0f239a/megatron/training.py#L783) - add a step where we retrieve the teacher's logits. In this case would we need to change the `deepspeed.PipelineEngine` internals?
2- Store the embedding layer of the teacher model inside the student model and gather the last hidden states of the teacher model. Once this is gathered apply the forward pass with the embedding layer of the teacher model to get the logits. (cc @thomasw21 as discussed offline)

# main TODOs

- [x] load teacher model from a checkpoint - make sure to use a copy of the chkpt
- [ ] load student model from a checkpoint - make sure to use a copy of the chkpt
- [x] make the broadcasting of the teacher logits (or last hidden states) work. 
- [x] try on 176

### CODE REVIEW

File: megatron/training.py [hunk 1]
```diff
@@ -366,6 +522,128 @@ def get_learning_rate_scheduler(optimizer):
...
+            assert teacher_model.grid.get_data_parallel_rank() == mpu.get_data_parallel_rank()
+        teacher_model = [teacher_model]
+    
+    # Load teacher model
+    timers = get_timers()
+    # Extra barrier is added to make sure all ranks report the
+    # max time.
+    torch.distributed.barrier()
+    timers('load-checkpoint').start()
+    # load_teacher_checkpoint(teacher_model)
```
Comment (younesbelkada) [inline megatron/training.py:633]: for now comment this, but it will be un-commented for the final script
